File: 143-reorder-list/reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    ListNode* front;
    bool stop;
    void reorder(ListNode* curr) {
        if (!curr) return;

        reorder(curr->next);

        if (stop) return;

        if (front == curr || front->next == curr) {
            curr->next = nullptr;
            stop = true;
            return;
        }

        ListNode* nextFront = front->next;
        front->next = curr;
        curr->next = nextFront;
        front = nextFront;
    }
public:
    void reorderList(ListNode* head) {
        front = head;
        stop = false;
        reorder(head);
    }
};

File: 143-reorder-list/reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode front;
    private boolean stop;
    private void reorder(ListNode curr){
        if (curr == null) return;

        reorder(curr.next);

        if (stop) return;

        if (front == curr || front.next == curr) {
            curr.next = null;
            stop = true;
            return;
        }

        ListNode nextFront = front.next;
        front.next = curr;
        curr.next = nextFront;
        front = nextFront;
    }
    public void reorderList(ListNode head) {
        stop = false;
        front = head;
        reorder(head);
    }
}